### PR TITLE
Rewrite consumer to support multiple partitions

### LINKED ISSUE #140: MultiConsumer
@wvanbergen @snormore @mkobetic (and anyone else who cares)

The current consumer works just fine, but is limited to fetching just a single partition of a single topic at once. You can run multiple consumers, but this has several issues:
 - They can't use the same client, or else the broker connection has contention problems (brokers handle requests on a single connection serially, so if it has to wait for `MaxWaitTime` before responding, this introduces severe latency when multiplexing busy and non-busy topics). Using separate clients works, but generates a *lot* of extra connections and metadata traffic.
 - You don't get messages from multiple partitions batched into a single response, when that makes sense for efficiency.

The ideal solution is to have a consumer capable of fetching from multiple topic/partitions at once. This has been at the back of my mind for a while now, so I already have a design ready; if the design makes sense to y'all, then somebody just needs to write it :)

While the user currently specifies a single topic/partition to the constructor, they should now be able to specify a set of topic/partitions. Since some of the configuration that is currently a singleton actually needs to be per-partition (`OffsetMethod` and `OffsetValue` at least), this would probably be of type `map[string]map[int32]*ConsumerPartitionConfig`. I considered permitting the dynamic adding/removing of partitions to the set, but I don't see a strong use case and it complicated a bunch of things.

Events are returned to the user exactly the same way they are now, over a single channel. I considered a separate channel per topic/partition but it complicated the base case, and the events already contain information on topic/partition so it's not hard for the user to dispatch appropriately if they really want to.

The constructor starts up one "controller" goroutine, which starts up and manages one goroutine per broker-that-has-a-partition-we-care-about and is responsible for (initially) dispatching each topic/partition to the appropriate broker's goroutine. The broker goroutine looks a lot like the current `fetchMessages` method with a few tweaks:
 - Some minor work needed to handle multiple blocks in the requests and responses.
 - When a topic/partition is reassigned to a new broker, that topic/partition gets returned to the controller via a channel; the goroutine tracks how many it is "responsible" for and exits if that reaches 0.
 - Similarly when a broker goes down hard, all topic/partitions are returned to the controller for re-dispatching and the goroutine exits.

I expect the success case to be fairly straightforward - as always, the complexity will come when reasoning through the failure cases and ensuring that topic/partitions are redispatched correctly, messages are not accidentally skipped in that case, etc. etc.

When the consumer is closed, it signals the controller which cleans up its children before exiting.

Thoughts?

### LINKED ISSUE #200: Consumer tests take forever
This has been bugging me forever, filing a ticket so I don't forget. These tests take anywhere from 0.5-10 seconds apparently at random, when they are entirely local and with appropriate configuration should basically be instantaneous. Something is fishy here.

cc @wvanbergen

### LINKED ISSUE #225: Handling of compacted topics
It appears that Sarama currently has problems on compacted topics. I have the follow test case(error handling omitted for brevity):
```go
package main

import (
  "github.com/Shopify/sarama"
  "log"
)

func main() {
  cfg := sarama.NewClientConfig()
  cli, _:= sarama.NewClient("", []string{"localhost:9092"}, cfg)
  ccfg := sarama.NewConsumerConfig()
  ccfg.OffsetMethod = sarama.OffsetMethodOldest
  consumer, _ := sarama.NewConsumer(cli, "compacted_topic", 0, "", ccfg)
  for m := range consumer.Events() {
    log.Printf("%s", m)
  }
}
```
I run this against a Kafka 0.8.1 broker, with the compacted_topic being a compacted topic with options:
```
segment.bytes=100000000
cleanup.policy=compact
segment.ms=3600000
```
The topic has a single partition, which currently starts at offset 169314.

When I run the above code, I receive a stream of:
```
2014/12/15 01:23:18 &{  expert.alerts %!s(int32=0) %!s(int64=0) kafka: Response did not contain all the expected topic/partition blocks.}
```
Which I believe is the IncompleteResponse error. After some digging around, it appears that Kafka reports the initial offset(using GetOffset) as 8, but when the actual request is made the offset is 169314. Kafkacat(via librdkafka) handles this gracefully, but sarama simply keeps giving that same error. Looking at the code, I believe the error is produced at:
https://github.com/Shopify/sarama/blob/master/consumer.go#L277-L285

I'm not sure if this a Sarama oversight, or a Kafka bug, but I believe that compaction can introduce gaps in the offset indexes received, so it seems like an oversight in Sarama. My gut tells me that the code needs to check that the offset is monotonically increasing, but can't assume that there are no gaps.

Best regards,
Lasse Dalegaard

### PR DESCRIPTION
**Pull request by eapache**
- `Consumer` is now a much lighter "manager" that spawns `PartitionConsumer`s
  instead. `PartitionConsumer`s behave and have an API almost exactly like the
  old `Consumer`.
- Automatically groups and redispatches PartitionConsumers when leadership moves
  in the cluster.

As with the producer-ng, this is effectively a rewrite and (with the exception of the changes to functional_test.go) should be reviewed wholesale rather than by diff.

@mkobetic @snormore @wvanbergen @yagnik @drdee and anyone else who cares, this is ready to be looked at whenever is convenient. Thoughts on the API? Thoughts on the TODOs still sprinkled about?

Fixes #140 
Fixes #200 
Fixes #225

### CODE REVIEW

**Comment by eapache** [comment]
As with the producer, I wrote a brief architectural wiki page which should help with review: https://github.com/Shopify/sarama/wiki/Consumer

**Comment by mkobetic** [comment]
In light of the discussions in https://github.com/Shopify/reportify/issues/1190, we're seriously considering bumping up the partition counts on the analytics topics. But for that we will definitely need this feature. Given that the stack is not in production yet, we may be a good guinea pigs for this. Any concerns about that @eapache?

**Comment by eapache** [comment]
Go for it - if you want an overview/architecture/review discussion we can schedule that whenever you want.

**Comment by wvanbergen** [comment]
:balloon:

**Comment by snormore** [comment]
:metal:

File: consumer.go [hunk 1]
```diff
@@ -89,298 +145,453 @@ func NewConsumer(client *Client, topic string, partition int32, group string, co
...
+	child.consumer.removeChild(child)
+	close(child.events)
+}
+
+func (child *PartitionConsumer) dispatch() error {
+	if err := child.consumer.client.RefreshTopicMetadata(child.topic); err != nil {
+		return err
+	}
+
+	if leader, err := child.consumer.client.Leader(child.topic, child.partition); err != nil {
```
Comment (mkobetic) [inline consumer.go]: Same chunk of code is in ConsumePartition, could it just call dispatch() ?

Comment (eapache) [inline consumer.go]: The problem is the `addChild()` in the middle of this code in `ConsumerPartition` - if I just call `dispatch()` then that call gets moved and things can happen in the wrong order. It's not obvious to me how to fix that without just duplicating the code.

Comment (mkobetic) [inline consumer.go]: Ah, I missed that. :+1:

File: consumer.go [hunk 2]
```diff
@@ -89,298 +145,453 @@ func NewConsumer(client *Client, topic string, partition int32, group string, co
...
+			w.abort(err)
+			return
+		}
+
+		for child, _ := range w.work {
+			block := response.GetBlock(child.topic, child.partition)
+			if block == nil {
+				// TODO should we be doing anything else here? The fact that we didn't get a block at all
+				// (not even one with an error) suggests that the broker is misbehaving, or perhaps something in the
+				// request/response pipeline is ending up malformed, so we could be better off aborting entirely...
```
Comment (mkobetic) [inline consumer.go]: Can we check whether the response has any valid blocks above?

Comment (eapache) [inline consumer.go]: What would that win us?

File: consumer.go [hunk 3]
```diff
@@ -89,298 +145,453 @@ func NewConsumer(client *Client, topic string, partition int32, group string, co
...
 
-	if config.MaxMessageSize < 0 {
-		return ConfigurationError("Invalid MaxMessageSize")
+	return w.broker.Fetch(w.consumer.client.id, request)
+}
+
+func (w *consumerWorker) handleResponse(child *PartitionConsumer, block *FetchResponseBlock) {
+	switch block.Err {
+	case NoError:
+		break
```
Comment (mkobetic) [inline consumer.go]: Why do we need the break here? Because there's nothing else in the clause?

Comment (eapache) [inline consumer.go]: I guess it's not strictly necessary, but it makes it clearer (to me, coming from C code) that this case doesn't fall through into the one below :)

File: consumer.go [hunk 4]
```diff
@@ -89,298 +145,453 @@ func NewConsumer(client *Client, topic string, partition int32, group string, co
...
+
+	atLeastOne := false
+	prelude := true
+	for _, msgBlock := range block.MsgSet.Messages {
+
+		for _, msg := range msgBlock.Messages() {
+			if prelude && msg.Offset < child.offset {
+				continue
+			}
+			prelude = false
```
Comment (mkobetic) [inline consumer.go]: prelude means that we've already seen this message?

File: consumer.go (hunk 2 above)
Comment (mkobetic) [inline consumer.go]: I was just thinking about the comment there, that maybe we should abort, but what if there are other valid blocks in the response. But it seems that right now we won't detect the case when there isn't any valid block in the response. Would we want to abort in that case?

File: consumer.go (hunk 4 above)
Comment (eapache) [inline consumer.go]: If we ask for offset X, kafka is allowed to send us a few messages <X if it already has e.g. a buffer in memory from X-5 to X+5, so we have to just ignore those messages. This happens in particular when using compression, since messages get compressed in batches and kafka is just going to send us the entire batch containing whichever offset we ask for.

File: consumer.go (hunk 3 above)
Comment (mkobetic) [inline consumer.go]: I agree it is clearer this way. I just wanted to make sure I'm not missing something important.

File: consumer.go (hunk 2 above)
Comment (eapache) [inline consumer.go]: Hmm, that makes sense... there are similar comments in `handleResponse`, I wonder if it will play nicely with that

**Comment by mkobetic** [comment]
So I went through the code, unsurprisingly it's pretty tight and well written. I admit not having entirely clear picture of how the various channels interact, so that would be something I'd love to cover in our review session on Friday. BTW, I recommend running this through godoc -analyze, I find it helps especially navigating the channels. Here's a sample illustration:

screen shot 2015-02-11 at 1 33 22 pm

File: consumer.go (hunk 2 above)
Comment (eapache) [inline consumer.go]: I just pushed https://github.com/Shopify/sarama/commit/9fdbfb71360174abee96d44bb19efec7d7e945f8 which should address this. The commit message explains my logic.

Comment (mkobetic) [inline consumer.go]: The rationale sounds good to me.

File: consumer.go [hunk 5]
```diff
@@ -8,43 +10,97 @@ import (
...
+}
+
+// PartitionConsumerConfig is used to pass multiple configuration options to AddPartition
+type PartitionConsumerConfig struct {
+	// The default (maximum) amount of data to fetch from the broker in each request. The default is 32768 bytes.
+	DefaultFetchSize int32
+	// The maximum permittable message size - messages larger than this will return MessageTooLarge. The default of 0 is
+	// treated as no limit.
+	MaxMessageSize int32
+	// The method used to determine at which offset to begin consuming messages. The default is to start at the most recent message.
```
Comment (drdee) [inline consumer.go]: Kafka docs warn that `fetch.message.max.bytes` should be at least as large as `message.max.bytes` else you can produce messages that cannot be consumed. 

Assuming that `DefaultFetchSize` maps to `fetch.message.max.bytes` and `MaxMessageSize` on the producer size is set to 1M bytes (https://github.com/Shopify/sarama/blob/afe4e78f8b33036dac4576e0a7f58ff9bcbedd9a/producer.go#L25)  then I think we should bump `DefaultFetchSize` from 32k to 1M bytes as well.

Comment (eapache) [inline consumer.go]: `MaxMessageSize` on the consumer maps to `fetch.message.max.bytes` and it defaults to unlimited.

**Comment by wvanbergen** [comment]
Updated the wiki pages to use the new names.

**Comment by wvanbergen** [comment]
One last naming issue: I am annoyed by the fact we have MessageToSend and ConsumerEvent. there's only a minor difference between them. Can they be merged? If not, can we at least both call them Message?

**Comment by wvanbergen** [comment]
Also, small inconsistency related to this: MessageToSend has no Err field. In case of an error, we wrap it in a ProducerError struct. For ConsumerEvent, we simply have a separate field in the struct that can be nil.

Maybe we should have a dual channel approach, one for events, and one for errors, like the Producer?

File: consumer.go (hunk 5 above)
Comment (drdee) [inline consumer.go]: What does `DefaultFetchSize` map to? And can't we have property names that map 1:1 to Kafka names?

Comment (eapache) [inline consumer.go]: > What does DefaultFetchSize map to?

Nothing, it's a feature kafka doesn't support.

> And can't we have property names that map 1:1 to Kafka names?

Our implementation is different (I intentionally ignored their consumer and didn't use it as a guide) because goroutines and channels lend themselves to a substantially different architecture, so using them to write a scala design would probably have gone poorly.

The producer has a similar "issue" - the scala implementation is much different under the hood, so not all the properties map exactly in sometimes subtle ways.

**Comment by eapache** [comment]
> One last naming issue: I am annoyed by the fact we have MessageToSend and ConsumerEvent. there's only a minor difference between them. 

The important difference (which isn't obviously surmountable) is that `MessageToSend` accepts encoders for the key/value, whereas `ConsumerEvent` can only return bytes. We could in theory have `MessageToSend` take bytes directly, but that requires the caller to serialize their payloads themselves, and it prevents them from using the actual golang values in e.g. partitioners if they so desire.

**Comment by eapache** [comment]
> If not, can we at least both call them Message?

They can't have the same name, but we could call them `ProducerMessage` and `ConsumerMessage` I guess...?

**Comment by wvanbergen** [comment]
+1 - it's a sucky backwards incompatible API change but it always bugs me :)

File: consumer.go (hunk 5 above)
Comment (drdee) [inline consumer.go]: Fair enough but this could be an ongoing source of confusion for people -- just like what I did they would look to the Kafka documentation and try to map Sarama's config options to the documented Kafka config  options. An idea to add as comments the Kafka counterpart config option (if it exists)?

**Comment by eapache** [comment]
Actually I kind of want to make the producer take bytes and really merge them:
- it avoids the need for the stupid `StringEncoder` (and most payloads seem to be strings or bytes already anyways)
- it deduplicates these two types
- smart partitioners can just use the fancy new `Metadata` field

**Comment by eapache** [comment]
But I'll do that in a separate PR

File: consumer.go (hunk 5 above)
Comment (eapache) [inline consumer.go]: I'd like to think that our documentation is good enough that they won't need to look at the kafka counterparts to understand - that is the first-order win imho

Comment (eapache) [inline consumer.go]: This can be a separate PR though

Comment (eapache) [inline consumer.go]: And to be clear, I'm not against adding links to counterparts, I just wish it weren't necessary :)

**Comment by wvanbergen** [comment]
All my ideas can be addressed in a separate PR. This can be merged IMHO.

**Comment by eapache** [comment]
I have opened tickets to track the larger changes coming out of this review. :rocket: time!

File: consumer.go (hunk 5 above)
Comment (wvanbergen) [inline consumer.go]: Any reason why `EventBufferSize`, `DefaultFetchSize` and `MaxMessageSize` are part of the Partition configuration, and `MinFetchSize` and `MaxWaitTime` of the Consumer configuration?

Personally I'd like them to be all part of the `ConsumerConfig`, so the `PartitionConsumerConfig` can just focus on offset management. Or use no `PartitionConsumerConfig` at all: `NewPartitionConsumer(consumer, topic, partition, startOffset)`, where startOffset can be -1 or -2 for the 2 special values.

Comment (eapache) [inline consumer.go]: Options available per-partition are because they are set per-partition in the kafka fetch request. Options available globally are because they are set globally in the kafka fetch request.

I agree it would be slightly nice to unify them, but I'm not sure how to do that without removing some of the flexibility that kafka provides.

Comment (wvanbergen) [inline consumer.go]: I wouldn't mind moving this to ConsumerConfig and lose some flexibility, but I guess it makes sense to stay close to the protocol.

I do think that separating out the offset parameter makes sense though, so in most cases you can just set the PartitionConsumerConfig parameter to `nil`.

``` go
// compare
consumer.ConsumePartition("topic", 1, sarama.OffsetOldest, nil)
consumer.ConsumePartition("topic", 1, 12345, nil)
// vs
config := sarama.NewPartitionConsumerConfig()
config.OffsetMethod = sarama.OffsetMethodManual
config.OffsetValue = 12345
consumer.ConsumePartition("topic", 1, config)
```

Comment (eapache) [inline consumer.go]: My concern is when we add new offset methods (offset-from-zookeeper or offset-from-kafka or whatever) then we will have to move it back to the config anyways in order to pass in the relevant information...
